sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/EliminateUnnecessaryOuterJoin.scala

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions.AttributeSet
+import org.apache.spark.sql.catalyst.plans.{LeftOuter, RightOuter}
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, Join, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * Removes outer join if it only has distinct on streamed side.
+ */
+object EliminateUnnecessaryOuterJoin extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case a @ Aggregate(_, _, p @ Project(_, Join(left, _, LeftOuter, _, _)))
```

Comment:
what if there is no Project between Aggregate and Join?

Reply:
It must has a `Project` between `Aggregate` and `Join` because their output is different if we can remove `Join`.


For example, we can not remove Join for the following case:
```sql
SELECT DISTINCT * FROM t1 LEFT JOIN t2 ON a = b

 Aggregate [a#2L, b#6L], [a#2L, b#6L]              
 +- Join LeftOuter, (a#2L = b#6L)                  
    :- Project [id#0L AS a#2L]                     
    :  +- Range (0, 200, step=1, splits=Some(2))   
    +- Project [id#4L AS b#6L]                     
       +- Range (0, 300, step=1, splits=Some(2))   
```

Reply:
why not? `Aggregate` defines its output and it won't change with the child.

Reply:
Because the supported case must the subset of the join's output(Either all on the left side or all on the right side):
![image](https://user-images.githubusercontent.com/5399861/112001045-871b5e00-8b59-11eb-9494-10a2cf035b0b.png)

Reply:
When this happens, column pruning will insert a Project between them. So I agree that the code here is correct in reality. But the code here assumes that column pruning will kick in first, and it's not necessary to have this assumption.

Reply:
ping @wangyum

Reply:
done